R/FilterStates.R

```diff
@@ -373,12 +373,18 @@ FilterStates <- R6::R6Class( # nolint
               logger::log_debug("FilterStates$srv_active@2 triggered by added states: { toString(added_state_names) }")
               lapply(added_states(), function(state) {
                 fs_callback <- state$server(id = fs_to_shiny_ns(state))
-                observeEvent(
+                state_id <- state$get_state()$id
+                remove_observer <- observeEvent(
                   once = TRUE, # remove button can be called once, should be destroyed afterwards
                   ignoreInit = TRUE, # ignoreInit: should not matter because we destroy the previous input set of the UI
                   eventExpr = fs_callback(), # when remove button is clicked in the FilterState ui
-                  handlerExpr = private$state_list_remove(state$get_state()$id)
+                  handlerExpr = private$state_list_remove(state_id)
                 )
+
+                # Keep track of all remove observers
+                remove_id_prefix <- sprintf("remove_%s", state_id)
+                remove_id <- sprintf("%s_%s", remove_id_prefix, sum(grepl(remove_id_prefix, names(private$observers))))
+                private$observers[[remove_id]] <- remove_observer
```

Comment:
For sake of simplicity, you can also:
- move `FilterState$<private>$observers` to `FilterState$<public>$observers`
- set this `remove_observer` as an element of `FilterState$observers` class
- then `state_list_remove` can remain unchanged and will call `state$destroy_observers()`